### ISSUE TITLE
Recent commits broke omnifun autocompletion with vimtex

### ISSUE THREAD

**Issue by clason** [ncm2/ncm2#22]
Before commit 22d80bc0b93c942eeb07d9c5cbd2ae6927d3cd64, the following setup correctly autocompleted:

### minimal vimrc

```viml
call plug#begin('~/.vim/plugged')
Plug 'roxma/nvim-yarp'
Plug 'ncm2/ncm2'
Plug 'lervag/vimtex'
call plug#end()

au BufEnter * call ncm2#enable_for_buffer()
au User Ncm2Plugin call ncm2#register_source({
            \ 'name' : 'vimtex',
            \ 'priority': 1,
            \ 'subscope_enable': 1,
            \ 'complete_length': 1,
            \ 'scope': ['tex'],
            \ 'mark': 'tex',
            \ 'word_pattern': '\w+',
            \ 'complete_pattern': g:vimtex#re#ncm,
            \ 'on_complete': ['ncm2#on_complete#omni','vimtex#complete#omnifunc'
],
            \ })
set completeopt=noinsert,menuone,noselect
```

### test files

`test.bib` (needs to be in the same directory)
```tex
@article{A1,
  author = {Author, A.},
  journal = {Good Journal},
  year = {2001},
  title = {Alpha},
}
@article{A2,
  author = {Author, A.},
  journal = {Very Good Journal},
  year = {2008},
  title = {Bravo},
}
@article{B3,
  author = {Buthor, C.},
  journal = {Also Good Journal},
  year = {2004},
  title = {Charlie},
}
```

`test.tex` (the main file)
```tex
\documentclass{article}
\begin{document}

\bibliographystyle{plain}
\bibliography{test.bib}
\end{document}
```
### steps to reproduce

**First problem:**

1. `nvim -u vimrc test.tex`
2. in the empty line after `\begin{document}`, type `\cit`

* expected: `ncm2` suggests `\cite` (same as `<c-x><c-o>`)
* observed: nothing happens

If I revert to 7feca385009bc62a844b8ffa98119edc7a2f8667, autocomplete works *here* as expected.

**Second problem:**

1. `nvim -u vimrc test.tex`
2. in the empty line after `\begin{document}`, type(!) `\cite{A`

Important: do not let `ncm2` complete the `\cite` command! Otherwise the issue won't occur.

* expected: `ncm2` suggests `A1` and `A2` (same as `<c-x><c-o>`)
* observed: nothing happens

Note: just typing `\cite{` does suggest `A1`, `A2`, `B1` as expected; it's just the further filtering that no longer works. If I revert to 0fd55917436284c429d9895ce2ac42e8902d3bf5, the filtering works as expected. 

### os

This happens with `neovim` release `0.3.0` as well as `v0.3.1-180-g487421413`, on macOS as well as Linux.

**Comment by roxma** [comment]
There're two cause of this issue.

## 1 Recent changes of the matcher behavior. 

The old behavior uses `complete_item['word']` for matching and filter.

The new behavior uses `complete_item['word']` for matching and filter.
But vimtex's `complete_item['abbr']` is quite different from `complete_item['word']`. Which doesn't work with the default prefix matcher..

In order to get things easier to work , I decided to change the default matcher to fuzzy matcher.

If you still want the prefix matcher works for vimtex, try

```vim
let ncm2#matcher = "prefix"
```

add add `"matcher": {"key": "word"}` when `register_source` for vimtex

## 2 A bug due to recent change

This is fixed by https://github.com/ncm2/ncm2/commit/84f8c178df172bb04940f9b8cba33101e2e8e45b

**Comment by roxma** [comment]
This example shows to use the prefix matcher for vimtex and leaving ncm2's default abbrfuzzy:

```vim
au User Ncm2Plugin call ncm2#register_source({
            \ 'name' : 'vimtex',
            \ 'priority': 1,
            \ 'subscope_enable': 1,
            \ 'complete_length': 1,
            \ 'scope': ['tex'],
            \ 'matcher': {'name': 'prefix', 'key': 'word'},
            \ 'mark': 'tex',
            \ 'word_pattern': '\w+',
            \ 'complete_pattern': g:vimtex#re#ncm,
            \ 'on_complete': ['ncm2#on_complete#omni', 'vimtex#complete#omnifunc'],
            \ })
```

The key is `           \ 'matcher': {'name': 'prefix', 'key': 'word'},`

**Comment by clason** [comment]
Yes, that works, perfect! I'll report the change to https://github.com/lervag/vimtex so the documentation can be updated. (Btw, you can link to the [documentation](https://github.com/lervag/vimtex/blob/master/doc/vimtex.txt) or https://github.com/lervag/vimtex/issues/1160 from the wiki. Also, the ncm2 documentation still lists the `swapcase_word` as the default sorter.)

I'm really impressed by your development speed! Thank you for your hard work!

**Comment by roxma** [comment]
@clason 

Thanks for the catch.

I have accidently added a debug comment in 84f8c17, which is removed in later version.

Please test the latest version.

**Comment by clason** [comment]
Still works!

**Comment by roxma** [comment]
Since the bib title is in the `menu` field, use `abbr` to filter the `menu` field might be convenient too...

`'matcher': {'name': 'abbrfuzzy', 'key': 'menu'}`

![rec](https://user-images.githubusercontent.com/4538941/42722931-1495be9e-8787-11e8-80e8-bef693b97155.gif)

**Comment by clason** [comment]
You're right, this would even better -- but now it no longer completes based on the author string?
